plutus-ir/src/Language/PlutusIR/Compiler/Datatype.hs

```diff
@@ -0,0 +1,294 @@
+-}
+
+{- Note [Abstract data types]
+While the Scott encoding makes it easy to simply convert all types inline, it
+is convenient for a number of reasons to instead abstract out data types. Namely:
+- The resulting code is much more readable, since we have named types instead
+of (potentially big) inlined types.
+- We generate less code because we're not repeating the definition at every use site.
+
+The basic idea is to "let-bind" types using type abstractions, and values using
+lambda abstractions. There are a few considerations that make this tricky, however:
+1. When types are inlined, the Scott encoding allows us to construct values by
+constructing the matching function inline. When they are abstract, we need to provide
+(suitably polymorphic) constructors abstractly.
+2. When types are inlined, the Scott encoding allows us to match against a type by
+simply using it as a function. When they are abstract, we need to provide a
+(suitably polymorphic) matching function abstractly.
+3. The definition of a type must be abstract in the binding for the constructors and destructors
+(so they can be used alongside the abstract type), but it must *not* be abstract in the
+*body* of the constructors or destructors, because they depend on knowing the real structure
+of the type.
+
+Consequently, for a single type we end up with something like the following:
+
+(abs ty :: (type) .
+  \c1 : forall t_1 :: (type) .. t_i :: (type) . t_1 -> .. -> t_c_1_k -> ty .. c_j : t_1 -> t_c_j_k -> ty
+    \match : forall t_1 :: (type) .. t_i :: (type) . forall r :: (type) b. (t_1 -> .. -> t_c_1_k -> r) -> .. -> (t_1 -> t_c_j_k -> r) -> r
+      <user term>
+)
+<defn of t>
+<defn of c1>
+..
+<defn of c_n>
+<defn of match>
+(see Note [Scott-encoding of datatypes] for how the actual definitions are constructed)
+
+We also have to convert data constructors and case expressions into suitably instantiated
+calls to the constructors/destructors instead of converting them directly as described in
+Note [Scott encoding of data types].
+-}
+
+-- See note [Scott encoding of datatypes]
+-- | Make the "Scott-encoded" type for a 'Datatype', with type variables free.
+-- @mkScottTy "Maybe" = "forall r. r -> (a -> r) -> r"@
+mkScottTy :: MonadQuote m => Datatype TyName Name () -> m (PLC.Type TyName ())
+mkScottTy (Datatype _ _ _ _ constrs) = do
+    -- TODO: include type name
+    resultType <- liftQuote $ freshTyName () "r"
+    let caseTys = fmap (constructorCaseType (PLC.TyVar () resultType)) constrs
+        match = mkIterTyFun caseTys (PLC.TyVar () resultType)
+        universal = PLC.TyForall () resultType (PLC.Type ()) match
+    pure universal
+
+-- | Make the "pattern functor" of a 'Datatype'. This is just the normal type, but with the
+-- type variable for the type itself free. In the case of non-recursive datatypes this is just the
+-- datatype.
+-- @mkDatatypePatternFunctor "List" = "\a :: * -> forall r . r -> (a -> list a -> r) -> r"@
```

Comment:
Comments like this are absolutely great. Are there any other places where we can have them?

Reply:
I only really did them here because what these functions do is simple enough to have a one-line example (and it's kind of a "fake" example anyway), but I'll see if I can put a few more in.